File: backend/src/main/kotlin/taskmanager/backend/models/Note.kt
package taskmanager.backend.models

import org.bson.types.ObjectId
import taskmanager.backend.dtos.response.NoteResponseDto
import taskmanager.backend.models.base.impl.BaseEntityImpl
import taskmanager.backend.models.interfaces.AttachedToTaskEntity
import taskmanager.backend.models.interfaces.MappableCreatedByUserEntity

data class Note(
    override val createdBy: ObjectId,
    override val task: ObjectId,

    var header: String,
    var text: String
) : BaseEntityImpl(), AttachedToTaskEntity, MappableCreatedByUserEntity<NoteResponseDto> {

    override fun toResponseDto(userId: ObjectId, createdBy: User?): NoteResponseDto {
        return NoteResponseDto(
            _id = _id.toString(),
            createdBy = createdBy?.toResponseDto(),
            task = task.toString(),
            header = header,
            text = text,
            createdAt = createdAt,
            updatedAt = updatedAt,
            canEdit = userId.toString() == userId.toString()
        )
    }
}

File: frontend/lib/models/note.dart
import 'package:frontend/models/base/base_entity.dart';
import 'package:frontend/models/user.dart';

class Note extends BaseEntity {
  User? createdBy;
  String task;
  String header;
  String text;
  bool canEdit;

  Note({
    super.id,
    this.createdBy,
    this.task = '',
    this.header = '',
    this.text = '',
    super.createdAt,
    super.updatedAt,
    this.canEdit = false,
  });

  factory Note.fromJson(Map<String, dynamic> json) {
    return Note(
      id: json['_id'],
      createdBy: json['createdBy'] != null ? User.fromJson(json['createdBy'] as Map<String, dynamic>) : null,
      task: json['task'],
      header: json['header'],
      text: json['text'],
      createdAt: BaseEntity.parseDateFromJson(json['createdAt']),
      updatedAt: BaseEntity.parseDateFromJson(json['updatedAt']),
      canEdit: json['canEdit'],
    );
  }
}
